src/jit/codegenlegacy.cpp

```diff
@@ -18880,32 +18899,78 @@ regMaskTP CodeGen::genCodeForCall(GenTreeCall* call, bool valUsed)
 
                     if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
                     {
-                        // MOV vptrReg, [REG_CALL_IND_SCRATCH + vtabOffsOfIndirection]
-                        getEmitter()->emitIns_R_AR(ins_Load(TYP_I_IMPL), EA_PTRSIZE, vptrReg, vptrReg,
-                                                   vtabOffsOfIndirection);
+                        if (isRelative)
+                        {
+#if defined(_TARGET_ARM_)
+                            unsigned offset = vtabOffsOfIndirection + vtabOffsAfterIndirection;
+
+                            // MOV vptrReg1, [REG_CALL_IND_SCRATCH + vtabOffsOfIndirection]
+                            getEmitter()->emitIns_R_R_I(ins_Load(TYP_I_IMPL), EA_PTRSIZE, vptrReg1, vptrReg,
+                                                        vtabOffsOfIndirection);
+
+                            if (unsigned_abs(offset) <= EMIT_ARM_MAX_IMM_ADD)
```

Comment:
This code checks whether offset can be encoded directly in ADD instruction, and if not - adds MOVW/MOVT for putting the constant to intermediate register. For some other instructions, this kind of logic is implemented in emitarm.cpp, which seems more suitable place for this. However, putting it there would require allocation of one more register in this place. Could you, please, suggest how is it better to solve this?